File: py/portfolio.py
from money import Money

class Portfolio:
    def __init__(self):
        self._exchange_rates = {
                "EUR->USD": 1.2,
                "USD->KRW": 1100
                }

    def add(self, *moneys):
        self.moneys = [ money for money in moneys ]

    def evaluate(self, bank, currency):
        sum = 0.0
        failures = []
        for money in self.moneys:
            try:
                sum += bank.convert(money, currency).amount
            except Exception as ex:
                failures.append(ex)
        if failures:
            failure_message = ", ".join(f.args[0] for f in failures)
            raise Exception("Missing exchange rate(s):[" + failure_message + "]")
        else:
            return Money(sum, currency)

File: js/portfolio.js
const { fail } = require("assert");
const Money = require("./money")

class Portfolio {
  constructor() {
    this.moneys = []
  }

  add(...moneys) {
    this.moneys = this.moneys.concat(moneys);
  }

  evaluate(bank, currency) {
    let failures = [];
    let total = this.moneys.reduce((sum, money) => {
      try {
        let convertedMoney = bank.convert(money, currency);
        return sum + convertedMoney.amount;
      }
      catch (error) {
        failures.push(error.message);
        return sum;
      }
    }, 0);
    if (failures.length) {
      throw new Error("Missing exchange rate(s): [" + failures.join() + "]");
    } else {  // else is required here because we don't want to return
      return new Money(total, currency);  // a Money in case of errors
    }
  }

}

module.exports = Portfolio;
